distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,782 @@
+            self._frame_nbytes_needed -= n_read
+
+    def _reset_default_buffer(self):
+        start = self._default_start
+        end = self._default_end
+
+        if start < end and start != 0:
+            self._default_buffer[: end - start] = self._default_buffer[start:end]
+            self._default_start = 0
+            self._default_end = end - start
+        elif start == end:
+            self._default_start = 0
+            self._default_end = 0
+
+    def _message_completed(self):
+        self._queue.put_nowait(self._frames)
+        self._nframes = None
+        self._frames = None
+        self._frame_lengths = None
+        self._frame_nbytes_remaining = 0
+
+    def connection_lost(self, exc=None):
+        if exc is None:
+            exc = CommClosedError("Connection closed")
+        self._exception = exc
+        self._transport = None
+        self._is_closed.set_result(None)
+
+        # Unblock read, if any
+        self._queue.put_nowait(_COMM_CLOSED)
+
+        # Unblock write, if any
+        if self._paused:
+            waiter = self._drain_waiter
+            if waiter is not None:
+                self._drain_waiter = None
+                if not waiter.done():
+                    waiter.set_exception(exc)
+
+    def pause_writing(self):
+        self._paused = True
+
+    def resume_writing(self):
+        self._paused = False
+
+        waiter = self._drain_waiter
+        if waiter is not None:
+            self._drain_waiter = None
+            if not waiter.done():
+                waiter.set_result(None)
+
+    async def read(self):
+        # Even if comm is closed, we still yield all received data before
+        # erroring
+        if self._queue is not None:
+            out = await self._queue.get()
+            if out is not _COMM_CLOSED:
+                return out
+            self._queue = None
+        raise self._exception
+
+    async def write(self, frames):
+        if self._exception:
+            raise self._exception
+
+        nframes = len(frames)
+        frames_nbytes = [nbytes(f) for f in frames]
+        # TODO: the old TCP comm included an extra `msg_nbytes` prefix that
+        # isn't really needed. We include it here for backwards compatibility,
+        # but this could be removed if we ever want to make another breaking
+        # change to the comms.
+        msg_nbytes = sum(frames_nbytes) + (nframes + 1) * 8
+        header = struct.pack(f"{nframes + 2}Q", msg_nbytes, nframes, *frames_nbytes)
+        frames = [header, *frames]
+
+        if msg_nbytes < 2 ** 17:  # 128kiB
+            # small enough, send in one go
+            frames = [b"".join(frames)]
+
+        if len(frames) > 1:
+            self._transport.writelines(frames)
+        else:
+            self._transport.write(frames[0])
+        if self._transport.is_closing():
+            await asyncio.sleep(0)
+        elif self._paused:
+            self._drain_waiter = self._loop.create_future()
```

Comment:
Why a Future instead of an `asyncio.Event` or something? Just because it's lighter-weight?

Reply:
Yeah, just lighter weight.